### PR TITLE
Improve handling of TF CUDA tests for 14_1_X

### PR DESCRIPTION
**Pull request by valsdav**
#### PR description:

This PR improves the handling of CUDA unit tests for the TensorFlow package, using a new `tf_cuda_support` tool from scram , which checks if the GPU support is enabled in TensorFlow compilation. 

The PR also makes the TF cuda tests more strict by checking explicitely if a CUDA device is visible to TF and not only to cmssw.  

The test `testTFVisibleDevicesCUDA` is in fact run by the framework as a CUDA device is registered, but then TF does not recognize the device and the test fails.   The other `testTF*CUDA` tests were passing silently using the CPU to run the test.  After this PR all the TF sessions using  `tf::backend::cuda` , but  not finding a GPU will fail explicitly. 

- This PR is needed to continue the integration of CUDA 12.4 in https://github.com/cms-sw/cmsdist/pull/9046. 
- This PR needs https://github.com/cms-sw/cmsdist/pull/9066 (thanks to @smuzaffar https://github.com/cms-sw/cmssw/pull/44376#issuecomment-1992693414)

### CODE REVIEW

**Comment by smuzaffar** [comment]
@valsdav , any idea why only `testTFVisibleDevicesCUDA` fails for GPU IBs (https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc12/CMSSW_14_1_GPU_X_2024-03-11-2300/unitTestLogs/PhysicsTools/TensorFlow#/89-89 ) ?

**Comment by valsdav** [comment]
> @valsdav , any idea why only `testTFVisibleDevicesCUDA` fails for GPU IBs (https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc12/CMSSW_14_1_GPU_X_2024-03-11-2300/unitTestLogs/PhysicsTools/TensorFlow#/89-89 ) ?

Hi @smuzaffar I realized this morning testing locally that only that test is failing because it is the only one explicitly checking the list of devices visible to TF.  

The other tests that are run when a CUDA device is visible to the cmssw framework are silently fall-backing to CPU running in TF.  I can add the checks of  the list of devices also in those tests to make them check CUDA usage explicitly.

**Comment by valsdav** [comment]
> <iftool name="tf_cuda_support">
>   all tests which requires TF to be build with GPU support
> </iftool>

That's  a good idea! Thanks!

File: PhysicsTools/TensorFlow/test/BuildFile.xml [hunk 1]
```diff
@@ -5,138 +5,138 @@
...
+<!--     <use name="FWCore/ParameterSet"/> -->
+<!--     <use name="FWCore/ParameterSetReader"/> -->
+<!--     <use name="FWCore/PluginManager"/> -->
+<!--     <use name="FWCore/ServiceRegistry"/> -->
+<!--     <use name="FWCore/Utilities"/> -->
+<!--     <use name="HeterogeneousCore/CUDAServices"/> -->
+<!--     <use name="HeterogeneousCore/CUDAUtilities"/> -->
+<!--     <use name="PhysicsTools/TensorFlow"/> -->
+<!--   </bin> -->
+<!-- </iftool> -->
```
Comment (fwyzard) [inline PhysicsTools/TensorFlow/test/BuildFile.xml:24]: if you want to comment out a whole test, you can just do
```suggestion
<!--
<iftool name="cuda">
  <bin name="testTFHelloWorldCUDA" file="testRunner.cpp,testHelloWorldCUDA.cc">
    <use name="boost_filesystem"/>
    <use name="catch2"/>
    <use name="cppunit"/>
    <use name="cuda"/>
    <use name="tensorflow-cc"/>
    <use name="FWCore/ParameterSet"/>
    <use name="FWCore/ParameterSetReader"/>
    <use name="FWCore/PluginManager"/>
    <use name="FWCore/ServiceRegistry"/>
    <use name="FWCore/Utilities"/>
    <use name="HeterogeneousCore/CUDAServices"/>
    <use name="HeterogeneousCore/CUDAUtilities"/>
    <use name="PhysicsTools/TensorFlow"/>
  </bin>
</iftool>
-->
```

**Comment by fwyzard** [comment]
> The other `testTF*CUDA` tests are instead silently using the CPU to run the test.

I think this part should be fixed: if we want the tests to run with CUDA, they should not fall back to CPU.

**Comment by fwyzard** [comment]
> I can add the checks of the list of devices also in those tests to make them check CUDA usage explicitly.

In addition to a check that there is a CUDA device available, can we actually force TF to run on the GPUs ?

**Comment by antoniovilela** [comment]
hold
   - As discussed at ORP.

**Comment by smuzaffar** [comment]
I will add `tf_cuda_support` tool and will redo this PR

**Comment by valsdav** [comment]
> > I can add the checks of the list of devices also in those tests to make them check CUDA usage explicitly.
> 
> In addition to a check that there is a CUDA device available, can we actually force TF to run on the GPUs ?

I have a solution for this, @smuzaffar do you want me to push it here so that it can be included in the new PR? Thanks

**Comment by smuzaffar** [comment]
@valsdav , I have opened https://github.com/cms-sw/cmsdist/pull/9066 which adds the suggested new tool `tf_cuda_support` . Feel free to update this PR with
- Changes needed for `In addition to a check that there is a CUDA device available, can we actually force TF to run on the GPUs`
- Move TF GPU unit tests in `<iftool name="tf_cuda_support">  </iftool>`  block

**Comment by valsdav** [comment]
please test with https://github.com/cms-sw/cmsdist/pull/9066

**Comment by smuzaffar** [comment]
enable gpu

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
@cms-sw/ml-l2 , can you please review this?
@cms-sw/orp-l2 , this now has all the required changes. you can `unhold`  it now

**Comment by valsdav** [comment]
+1

**Comment by smuzaffar** [comment]
@valsdav , can you please backport it for 14.0.X ?

**Comment by antoniovilela** [comment]
unhold

**Comment by antoniovilela** [comment]
+1

**Pull request merged**